pyomo/contrib/pynumero/algorithms/solvers/cyipopt_solver.py

```diff
@@ -0,0 +1,299 @@
+        self._nlp = nlp
+        self._is_composite = False
+        if hasattr(nlp, 'nblocks'):
+            self._is_composite = True
+
+        x = nlp.x_init()
+        y = nlp.y_init()
+        y.fill(1.0)
+
+        # get structures
+        self._df = nlp.grad_objective(x)
+        self._g = nlp.evaluate_g(x)
+        if not self._is_composite:
+            self._jac_g = nlp.jacobian_g(x)
+            self._hess_lag = nlp.hessian_lag(x, y)
+            self._hess_lower_mask = self._hess_lag.row >= self._hess_lag.col
+        else:
+            self._jac_g = nlp.jacobian_g(x)
+            expanded = self._jac_g.tocoo()
+            self._jac_row = expanded.row
+            self._jac_col = expanded.col
+
+            self._hess_lag = nlp.hessian_lag(x, y)
+            expanded = self._hess_lag.tocoo()
+            self._hess_lower_mask = expanded.row >= expanded.col
+            self._hess_row = np.compress(self._hess_lower_mask, expanded.row)
+            self._hess_col = np.compress(self._hess_lower_mask, expanded.col)
+
+    def objective(self, x):
+        return self._nlp.objective(x)
+
+    def gradient(self, x):
+        self._nlp.grad_objective(x, out=self._df)
+        if not self._is_composite:
+            return self._df
+        return self._df.flatten()
+
+    def constraints(self, x):
+        self._nlp.evaluate_g(x, out=self._g)
+        if not self._is_composite:
+            return self._g
+        return self._g.flatten()
+
+    def jacobian(self, x):
+        self._nlp.jacobian_g(x, out=self._jac_g)
+        if not self._is_composite:
+            return self._jac_g.data
+        return self._jac_g.coo_data()
+
+    def hessianstructure(self):
+        if not self._is_composite:
+            row = np.compress(self._hess_lower_mask, self._hess_lag.row)
+            col = np.compress(self._hess_lower_mask, self._hess_lag.col)
+            return row, col
+        return self._hess_row, self._hess_col
+
+    def jacobianstructure(self):
+        if not self._is_composite:
+            return self._jac_g.row, self._jac_g.col
+        return self._jac_row, self._jac_col
+
+    def hessian(self, x, lagrange, obj_factor):
+        self._nlp.hessian_lag(x,
+                              lagrange,
+                              out=self._hess_lag,
+                              eval_f_c=False,
+                              obj_factor=obj_factor)
+        if not self._is_composite:
+            data = np.compress(self._hess_lower_mask, self._hess_lag.data)
+            return data
+        data = np.compress(self._hess_lower_mask, self._hess_lag.coo_data())
+        return data
+
+    def intermediate(
+            self,
+            alg_mod,
+            iter_count,
+            obj_value,
+            inf_pr,
+            inf_du,
+            mu,
+            d_norm,
+            regularization_size,
+            alpha_du,
+            alpha_pr,
+            ls_trials
+    ):
+        pass
+
+
+class IpoptSolver(object):
```

Comment:
Several questions about this class:
- Why is this not named `CyIpoptSolver`?  It is specific to that solver interface...
- Why is this not a Pyomo solver?  Or would it be better to make a formal Pyomo Solver that takes a Pyomo model, creates the PyomoNLP, passes it to an instance of this class and calls `solve()`?

Reply:
Right, I will rename this to CyIpoptSolver. 

I agree that at some point CyIpoptSolver should be a formal pyomo solver. For now I have been mostly experimenting with it. I suggest moving CyIpoptSolver to a formal solver in a future pull request. However, I don't think there will be much benefit in creating the PyomoNLP at the solve( ) call. The main advantage of this approach is that users may modified attributes of Pynumero NLPs without rewriting the NL-file. For instance, I have implemented ADMM with this and \rho and \lambda_i can be updated directly in the NLP interface avoiding rewriting the NL file at every NLP iteration. Another good example is solving a TwoStageStochasticNLP that is a composite NLP built pyomo models.

Reply:
I will be happy to discuss this further and improve the design of CyIpoptSolver. In particular because it helps a lot for external-decomposition approaches. It can also help towards a persistent interface to NLPs

Reply:
Renamed IpoptSolver to CyIpoptSolver